lib/uklock: adding mutex metrics

RaduMantu opened the pull request:
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://docs.unikraft.org/contribute.html) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`
 - Platform(s): `kvm`
 - Application(s): N/A

### Description of changes

Added metric collection for `uklock/mutex`. These include active (un)locked mutex objects and total (un)successful (un)locks since startup.

Based on Cezar's branch: [ukstore_fixed_tree](https://github.com/craciunoiuc/unikraft/tree/ukstore_fixed_tree).

nderjung commented:
Hi @RaduMantu, thanks for submitting this PR and for the great work at [USoC21](https://usoc21.unikraft.org)!

To get this PR ready to be merged, please can you:

 * Remove the `ukstore` and `uktest` base commits; these libraries will be introduced by separate PRs (commits 0d2aa9e through 8ef440e need to be removed);
  * Please can you prefix all commit messages with `uklock: `;
  *  Please check the results of the `checkpatch.pl` program; looks like it failed.
  
  Thanks again!

RaduMantu commented:
Hi. Again, sorry for the delay.

I removed Cezar's commits and squashed / reworded mine. `checkpatch` should't cause any issues but I'll follow up once `ukstore` and `uktest` are integrated.

craciunoiuc commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -57,6 +60,31 @@ struct uk_mutex {
+/*
+ * Metric storage (see mutex.c).
+ */
+extern struct uk_mutex_metrics mutex_metrics;
+extern spinlock_t              mutex_metrics_lock;
```
I would prefix these like (and their other appearances, of course):
```
mutex_metrics;        -> _uk_mutex_metrics;
mutex_metrics_lock;   -> _uk_mutex_metrics_lock;
```

This is to ensure that there won't be name conflicts in the future (someone might create another `mutex_metrics` symbol).

craciunoiuc commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -43,6 +43,9 @@
 #include <uk/wait_types.h>
 #include <uk/plat/time.h>
 
+#include <uk/arch/spinlock.h>
+#include <string.h>
```
Those might be worth protecting with `#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS` guards

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,69 @@
 
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+#include <uk/init.h>
+#include <uk/store.h>
```
Remove `#include <uk/store.h>`. The linking to `ukstore` can be introduced as another PR

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,69 @@
+}
+
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+/**
+ * Retrieves mutex-related metrics.
+ * @param cookie Unused.
+ * @param dst Reference to metrics copy buffer (allocated here).
+ * @return 0 on success, !0 otherwise.
+ *
+ * NOTE: this function is registered as a ukstore static store's getter
+ */
+static int uk_mutex_metrics_get(void *cookie __unused, __uptr *dst)
+{
+	/* allocate memory for copy of stats */
+	struct uk_mutex_metrics *to_send =
+		malloc(sizeof(struct uk_mutex_metrics));
+	if (!to_send) {
+		uk_pr_err("Unable to allocate memory for metrics structure\n");
+		return 1;
+	}
+
+	/* create copy */
+	memmove(dst, &mutex_metrics, sizeof(mutex_metrics));
+	*dst = (__uptr) to_send;
+
+	return 0;
 }
+UK_STORE_STATIC_ENTRY(mutex_stats, uptr, uk_mutex_metrics_get, NULL);
```
Also, remove this, as it is related to `ukstore`. It can be introduced back as a separate `ukstore` PR in the future.

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,69 @@
+
+/**
+ * Initializes the mutex metric counters and its spinlock.
+ * @return 0 on success, !0 otherwise.
```
It says it returns !0 on fail, but I only see a `return 0;`.

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,69 @@
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+#include <uk/init.h>
+#include <uk/store.h>
+#include <uk/print.h>
+#include <stdlib.h>
```
Also remove these 2. They are only needed in the `ukstore` context.

craciunoiuc reviewed (changes requested):
Hey @RaduMantu,

Thank you for your contributions.
I left you some comments in this review regarding your PR. Please address them before it can be merged.

Some additional notes regarding the PR:
 * In the commit description `currently` -> `Currently`
 * In the commit title `uklock:` -> `lib/uklock:`
 * Please change the PR title, as it won't be directly related to `ukstore` (it should be independent of it)
 * Suggestion: I would split this commit into 3:
   1. Registers the option inside `Config.uk`
   2. Registers the ctor with the init function & adds the new symbols
   3. Implements the decrementing/incrementing of fields

I was able to run run this PR and it worked as intended, functionality-wise there are no problems 🙂 

Cezar

RaduMantu committed:
lib/uklock: Add mutex metrics option to Config.uk

Signed-off-by: Radu Mantu <andru.mantu@gmail.com>

RaduMantu committed:
lib/uklock: Add mutex metrics ctor & register new symbols

Signed-off-by: Radu Mantu <andru.mantu@gmail.com>

RaduMantu committed:
lib/uklock: Add mutex metrics counters & definitions

Currently counting (un)locked objects and total (un)successful
(un)locks since startup.

Signed-off-by: Radu Mantu <andru.mantu@gmail.com>

RaduMantu commented:
Hey @craciunoiuc,

I've addressed the problems that you pointed out in your last review. Thanks for the feedback, btw.
Let me know if I missed anything.

Edit: added small fix to commit [5b0cd43](https://github.com/unikraft/unikraft/pull/282/commits/c8d9d6dacc646f3e9e58993544d00b1ef1c84506).

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,39 @@
+#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
+#include <uk/init.h>
+
+struct uk_mutex_metrics _uk_mutex_metrics;
```
```suggestion
struct uk_mutex_metrics _uk_mutex_metrics = { 0 };
```

Easily initialize everything to 0 so there is no need for `memset`

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,39 @@
+ */
+static int mutex_metrics_ctor(void)
+{
+	memset(&_uk_mutex_metrics, 0, sizeof(_uk_mutex_metrics));
```
```suggestion
```

You can remove this, as the structure is initialized to 0

craciunoiuc commented on lib/uklock/exportsyms.uk:
```diff
@@ -1,2 +1,4 @@
 uk_semaphore_init
 uk_mutex_init
+_uk_mutex_metrics
```
```suggestion
_uk_mutex_metrics
uk_mutex_get_metrics
```

Add a function to retrieve the structure with metrics

craciunoiuc commented on lib/uklock/mutex.c:
```diff
@@ -1,8 +1,39 @@
+	ukarch_spin_unlock(&_uk_mutex_metrics_lock);
+#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
+}
+
```
```suggestion
void uk_mutex_get_metrics(struct uk_mutex_metrics *dst)
{
	UK_ASSERT(dst);

	ukarch_spin_lock(&_uk_mutex_metrics_lock);
	memcpy(dst, &_uk_mutex_metrics, sizeof(*dst));
	ukarch_spin_unlock(&_uk_mutex_metrics_lock);
}
```

A separate function can do the retrieval so users do not use the `_uk_mutex_metrics` directly.

craciunoiuc replied:
You might need to include something `#include <uk/assert.h>`

craciunoiuc commented on lib/uklock/include/uk/mutex.h:
```diff
@@ -116,11 +168,25 @@ static inline void uk_mutex_unlock(struct uk_mutex *m)
+
 	ukplat_lcpu_restore_irqf(irqf);
 }
 
```
```suggestion
#ifdef CONFIG_LIBUKLOCK_MUTEX_METRICS
void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
#endif /* CONFIG_LIBUKLOCK_MUTEX_METRICS */
```

craciunoiuc reviewed (changes requested):
Hey @RaduMantu,

This should be the last round before this PR is approved. I had a few more ideas and added them as suggestions. Feel free to move them around to better places.

Cezar

RaduMantu committed:
lib/uklock: Add mutex metrics getter to avoid direct access

Signed-off-by: Radu Mantu <andru.mantu@gmail.com>

RaduMantu commented:
Thanks for the suggestions, @craciunoiuc.

craciunoiuc commented:
All good now. I tested it and it works.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

craciunoiuc reviewed (approved):
All good now. I tested it and it works.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

RaduMantu commented:
Fixed some `checkpatch` warnings. My bad.

craciunoiuc commented:
I guess I rushed with the approval before checkpatch. Sorry

craciunoiuc reviewed (approved):
All good now. I tested it and it works.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

nderjung reviewed (approved):
Great! Thanks 🐒 

Approved-by: Alexander Jung <a.jung@lancs.ac.uk>
